src/metrics/mod.rs

```diff
@@ -1,3 +1,116 @@
-pub mod crap;
-pub mod skunk;
-pub mod wcc;
+pub(crate) mod crap;
+pub(crate) mod skunk;
+pub(crate) mod wcc;
+
+use std::path::Path;
+
+use rust_code_analysis::{get_function_spaces, guess_language, read_file, FuncSpace};
+use serde::Serialize;
+use tracing::debug;
+
+use crate::{
+    error::{Error, Result},
+    Complexity, Thresholds,
+};
+
+const COVERAGE_THRESHOLD: f64 = 0.6;
+
+#[derive(Debug, Serialize, Clone, Copy)]
+#[serde(rename_all = "camelCase")]
+pub(crate) struct MetricsThresholds {
+    wcc: f64,
+    crap_cyclomatic: f64,
+    crap_cognitive: f64,
+    skunk_cyclomatic: f64,
+    skunk_cognitive: f64,
+}
+
+impl From<Thresholds> for MetricsThresholds {
+    fn from(value: Thresholds) -> Self {
+        MetricsThresholds {
+            wcc: value.wcc,
+            crap_cyclomatic: crap::crap(COVERAGE_THRESHOLD, value.cyclomatic_complexity),
+            crap_cognitive: crap::crap(COVERAGE_THRESHOLD, value.cognitive_complexity),
+            skunk_cyclomatic: skunk::skunk(COVERAGE_THRESHOLD, value.cyclomatic_complexity),
+            skunk_cognitive: skunk::skunk(COVERAGE_THRESHOLD, value.cognitive_complexity),
+        }
+    }
+}
+
+impl Default for MetricsThresholds {
+    fn default() -> Self {
+        Self {
+            wcc: 60.0,
+            crap_cyclomatic: 16.4,
+            crap_cognitive: 16.4,
+            skunk_cyclomatic: 16.66,
+            skunk_cognitive: 16.66,
+        }
+    }
+}
+
+impl MetricsThresholds {
+    pub(crate) fn is_complex(
+        &self,
+        wcc: f64,
+        crap: f64,
+        skunk: f64,
+        complexity: Complexity,
+    ) -> bool {
+        let (crap_threshold, skunk_threshold) = match complexity {
+            Complexity::Cyclomatic => (self.crap_cyclomatic, self.skunk_cyclomatic),
+            Complexity::Cognitive => (self.crap_cognitive, self.skunk_cognitive),
+        };
+
+        wcc < self.wcc || crap > crap_threshold || skunk > skunk_threshold
+    }
+}
+
+// Get the root FuncSpace from a file.
+#[inline]
+pub(crate) fn get_root(path: &Path) -> Result<FuncSpace> {
+    let source_code = read_file(path)?;
+    let language = guess_language(&source_code, path)
+        .0
+        .ok_or(Error::Language)?;
+
+    debug!("{:?} is written in {:?}", path, language);
+
+    let root = get_function_spaces(&language, source_code, path, None).ok_or(Error::Metrics)?;
+
+    Ok(root)
```

Comment:
Is there a specific error you are referring to? Because I took a look but the only exposed error in the `rust-code-analysis` API seems to be `ConcurrentErrors`, which I don't know if it makes sense here.

Reply:
You're right, we were adding new errors kinds, but we have not released anything yet. Let's leave as it is now.